File: multiple-languages/java/ros-cdk-cxapi/src/main/java/com/aliyun/ros/cdk/cxapi/CloudArtifact.java
package com.aliyun.ros.cdk.cxapi;

/**
 * Represents an artifact within a cloud assembly.
 */
@javax.annotation.Generated(value = "jsii-pacmak/1.85.0 (build 08ee592)", date = "2023-09-25T10:01:43.877Z")
@software.amazon.jsii.Jsii(module = com.aliyun.ros.cdk.cxapi.$Module.class, fqn = "@alicloud/ros-cdk-cxapi.CloudArtifact")
public class CloudArtifact extends software.amazon.jsii.JsiiObject {

    protected CloudArtifact(final software.amazon.jsii.JsiiObjectRef objRef) {
        super(objRef);
    }

    protected CloudArtifact(final software.amazon.jsii.JsiiObject.InitializationMode initializationMode) {
        super(initializationMode);
    }

    protected CloudArtifact(final @org.jetbrains.annotations.NotNull com.aliyun.ros.cdk.cxapi.CloudAssembly assembly, final @org.jetbrains.annotations.NotNull java.lang.String id, final @org.jetbrains.annotations.NotNull com.aliyun.ros.cdk.assembly.schema.ArtifactManifest manifest) {
        super(software.amazon.jsii.JsiiObject.InitializationMode.JSII);
        software.amazon.jsii.JsiiEngine.getInstance().createNewObject(this, new Object[] { java.util.Objects.requireNonNull(assembly, "assembly is required"), java.util.Objects.requireNonNull(id, "id is required"), java.util.Objects.requireNonNull(manifest, "manifest is required") });
    }

    /**
     * Returns a subclass of <code>CloudArtifact</code> based on the artifact type defined in the artifact manifest.
     * <p>
     * @return the <code>CloudArtifact</code> that matches the artifact type or <code>undefined</code> if it's an artifact type that is unrecognized by this module.
     * @param assembly The cloud assembly from which to load the artifact. This parameter is required.
     * @param id The artifact ID. This parameter is required.
     * @param artifact The artifact manifest. This parameter is required.
     */
    public static @org.jetbrains.annotations.Nullable com.aliyun.ros.cdk.cxapi.CloudArtifact fromManifest(final @org.jetbrains.annotations.NotNull com.aliyun.ros.cdk.cxapi.CloudAssembly assembly, final @org.jetbrains.annotations.NotNull java.lang.String id, final @org.jetbrains.annotations.NotNull com.aliyun.ros.cdk.assembly.schema.ArtifactManifest artifact) {
        return software.amazon.jsii.JsiiObject.jsiiStaticCall(com.aliyun.ros.cdk.cxapi.CloudArtifact.class, "fromManifest", software.amazon.jsii.NativeType.forClass(com.aliyun.ros.cdk.cxapi.CloudArtifact.class), new Object[] { java.util.Objects.requireNonNull(assembly, "assembly is required"), java.util.Objects.requireNonNull(id, "id is required"), java.util.Objects.requireNonNull(artifact, "artifact is required") });
    }

    /**
     * @return all the metadata entries of a specific type in this artifact.
     * @param type This parameter is required.
     */
    public @org.jetbrains.annotations.NotNull java.util.List<com.aliyun.ros.cdk.cxapi.MetadataEntryResult> findMetadataByType(final @org.jetbrains.annotations.NotNull java.lang.String type) {
        return java.util.Collections.unmodifiableList(software.amazon.jsii.Kernel.call(this, "findMetadataByType", software.amazon.jsii.NativeType.listOf(software.amazon.jsii.NativeType.forClass(com.aliyun.ros.cdk.cxapi.MetadataEntryResult.class)), new Object[] { java.util.Objects.requireNonNull(type, "type is required") }));
    }

    public @org.jetbrains.annotations.NotNull com.aliyun.ros.cdk.cxapi.CloudAssembly getAssembly() {
        return software.amazon.jsii.Kernel.get(this, "assembly", software.amazon.jsii.NativeType.forClass(com.aliyun.ros.cdk.cxapi.CloudAssembly.class));
    }

    public @org.jetbrains.annotations.NotNull java.lang.String getId() {
        return software.amazon.jsii.Kernel.get(this, "id", software.amazon.jsii.NativeType.forClass(java.lang.String.class));
    }

    /**
     * The artifact's manifest.
     */
    public @org.jetbrains.annotations.NotNull com.aliyun.ros.cdk.assembly.schema.ArtifactManifest getManifest() {
        return software.amazon.jsii.Kernel.get(this, "manifest", software.amazon.jsii.NativeType.forClass(com.aliyun.ros.cdk.assembly.schema.ArtifactManifest.class));
    }

    /**
     * The set of messages extracted from the artifact's metadata.
     */
    public @org.jetbrains.annotations.NotNull java.util.List<com.aliyun.ros.cdk.cxapi.SynthesisMessage> getMessages() {
        return java.util.Collections.unmodifiableList(software.amazon.jsii.Kernel.get(this, "messages", software.amazon.jsii.NativeType.listOf(software.amazon.jsii.NativeType.forClass(com.aliyun.ros.cdk.cxapi.SynthesisMessage.class))));
    }

    /**
     * Returns all the artifacts that this artifact depends on.
     */
    public @org.jetbrains.annotations.Nullable java.util.List<com.aliyun.ros.cdk.cxapi.CloudArtifact> getDependencies() {
        return java.util.Optional.ofNullable((java.util.List<com.aliyun.ros.cdk.cxapi.CloudArtifact>)(software.amazon.jsii.Kernel.get(this, "dependencies", software.amazon.jsii.NativeType.listOf(software.amazon.jsii.NativeType.forClass(com.aliyun.ros.cdk.cxapi.CloudArtifact.class))))).map(java.util.Collections::unmodifiableList).orElse(null);
    }
}

File: packages/@alicloud/ros-cdk-cxapi/lib/cloud-artifact.ts
import * as cxschema from "@alicloud/ros-cdk-assembly-schema";
import { CloudAssembly } from "./cloud-assembly";
import {
  MetadataEntryResult,
  SynthesisMessage,
  SynthesisMessageLevel,
} from "./metadata";

/**
 * Artifact properties for ROS stacks.
 */
export interface AliyunRosStackProperties {
  /**
   * A file relative to the assembly root which contains the ROS template for this stack.
   */
  readonly templateFile: string;

  /**
   * Values for ROS stack parameters that should be passed when the stack is deployed.
   */
  readonly parameters?: { [id: string]: string };

  /**
   * The name to use for the ROS stack.
   * @default - name derived from artifact ID
   */
  readonly stackName?: string;
}

/**
 * Represents an artifact within a cloud assembly.
 */
export class CloudArtifact {
  /**
   * Returns a subclass of `CloudArtifact` based on the artifact type defined in the artifact manifest.
   * @param assembly The cloud assembly from which to load the artifact
   * @param id The artifact ID
   * @param artifact The artifact manifest
   * @returns the `CloudArtifact` that matches the artifact type or `undefined` if it's an artifact type that is unrecognized by this module.
   */
  public static fromManifest(
    assembly: CloudAssembly,
    id: string,
    artifact: cxschema.ArtifactManifest
  ): CloudArtifact | undefined {
    // Implementation is defined in a separate file to break cyclic dependencies
    void(assembly), void(id), void(artifact);
    throw new Error('Implementation not overridden yet');
  }

  /**
   * The artifact's manifest
   */
  public readonly manifest: cxschema.ArtifactManifest;

  /**
   * The set of messages extracted from the artifact's metadata.
   */
  public readonly messages: SynthesisMessage[];

  /**
   * IDs of all dependencies. Used when topologically sorting the artifacts within the cloud assembly.
   * @internal
   */
  public readonly _dependencyIDs: string[];

  /**
   * Cache of resolved dependencies.
   */
  private _deps?: CloudArtifact[];

  protected constructor(
    public readonly assembly: CloudAssembly,
    public readonly id: string,
    manifest: cxschema.ArtifactManifest
  ) {
    this.manifest = manifest;
    this.messages = this.renderMessages();
    this._dependencyIDs = manifest.dependencies || [];
  }

  /**
   * Returns all the artifacts that this artifact depends on.
   */
  public get dependencies(): CloudArtifact[] | undefined {
    // if (this._deps) {
    //   return this._deps;
    // }

    // this._deps = this._dependencyIDs.map((id) => {
    //   const dep = this.assembly.tryGetArtifact(id);
    //   if (!dep) {
    //     throw new Error(`Artifact ${this.id} depends on non-existing artifact ${id}`);
    //   }
    //   return dep;
    // });

    return this._deps;
  }

  /**
   * @returns all the metadata entries of a specific type in this artifact.
   * @param type
   */
  public findMetadataByType(type: string): MetadataEntryResult[] {
    const result = new Array<MetadataEntryResult>();
    for (const path of Object.keys(this.manifest.metadata || {})) {
      for (const entry of (this.manifest.metadata || {})[path]) {
        if (entry.type === type) {
          result.push({ path, ...entry });
        }
      }
    }
    return result;
  }

  private renderMessages() {
    const messages = new Array<SynthesisMessage>();

    for (const [id, metadata] of Object.entries(this.manifest.metadata || {})) {
      for (const entry of metadata) {
        let level: SynthesisMessageLevel;
        switch (entry.type) {
          case cxschema.ArtifactMetadataEntryType.WARN:
            level = SynthesisMessageLevel.WARNING;
            break;
          case cxschema.ArtifactMetadataEntryType.ERROR:
            level = SynthesisMessageLevel.ERROR;
            break;
          case cxschema.ArtifactMetadataEntryType.INFO:
            level = SynthesisMessageLevel.INFO;
            break;
          default:
            continue;
        }

        messages.push({ level, entry, id });
      }
    }

    return messages;
  }
}
